adapters/bidmachine/bidmachine.go

```diff
@@ -0,0 +1,223 @@
+package bidmachine
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"text/template"
+
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/macros"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type BidmachineAdapter struct {
+	endpoint template.Template
+}
+
+func (a *BidmachineAdapter) MakeRequests(request *openrtb.BidRequest, _ *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("x-openrtb-version", "2.5")
+
+	impressions := request.Imp
+	result := make([]*adapters.RequestData, 0, len(impressions))
+	errs := make([]error, 0, len(impressions))
+
+	for i, impression := range impressions {
+		if impression.Banner == nil && impression.Video == nil {
+			errs = append(errs, &errortypes.BadInput{
+				Message: "Bidmachine supports only banner or video ads",
+			})
+			continue
+		}
+
+		if impression.Banner != nil {
+			banner := impression.Banner
+			if banner.Format == nil {
+				errs = append(errs, &errortypes.BadInput{
+					Message: "banner format required",
+				})
+				continue
+			}
+			if len(banner.Format) == 0 {
+				errs = append(errs, &errortypes.BadInput{
+					Message: "banner format array is empty",
+				})
+				continue
+			}
+		}
+
+		if len(impression.Ext) == 0 {
+			errs = append(errs, errors.New("impression extensions required"))
+			continue
+		}
+		var bidderExt adapters.ExtImpBidder
+		err := json.Unmarshal(impression.Ext, &bidderExt)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		if len(bidderExt.Bidder) == 0 {
+			errs = append(errs, errors.New("bidder required"))
+			continue
+		}
+		var impressionExt openrtb_ext.ExtImpBidmachine
+		err = json.Unmarshal(bidderExt.Bidder, &impressionExt)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		url, err := a.buildEndpointURL(impressionExt)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		if impressionExt.SellerID == "" {
+			errs = append(errs, errors.New("Bidmachine seller_id is required"))
+			continue
+		}
+		if impressionExt.Path == "" {
+			errs = append(errs, errors.New("Bidmachine path is required"))
+			continue
+		}
+		if bidderExt.Prebid != nil && bidderExt.Prebid.IsRewardedInventory == 1 {
+			if impression.Banner != nil && !hasRewardedBattr(impression.Banner.BAttr) {
+				impression.Banner.BAttr = append(impression.Banner.BAttr, openrtb.CreativeAttribute(16))
+
+			}
+			if impression.Video != nil && !hasRewardedBattr(impression.Video.BAttr) {
+				impression.Video.BAttr = append(impression.Video.BAttr, openrtb.CreativeAttribute(16))
+			}
+		}
+		request.Imp = impressions[i : i+1]
```

Comment:
With the shared memory fixes above, this will no longer work. You need to reference to impression copy made in this loop iteration:
```
request.Imp = []openrtb.Imp{impression}
```